File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/methods/userInteraction/AccountUsernameMethodHandler.kt
/*
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react.methods.userInteraction

import ch.nevis.mobile.authentication.sdk.react.MethodChannelHandler
import ch.nevis.mobile.authentication.sdk.react.cache.OperationCache
import ch.nevis.mobile.authentication.sdk.react.cache.operation.UserInteractionOperation
import ch.nevis.mobile.authentication.sdk.react.cache.state.SelectAccount
import ch.nevis.mobile.authentication.sdk.react.methods.MethodHandler
import ch.nevis.mobile.authentication.sdk.react.model.ReactMethodName
import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.AccountUsernameMessage
import ch.nevis.mobile.authentication.sdk.react.model.message.incoming.ChannelInMessage
import ch.nevis.mobile.sdk.api.MobileAuthenticationClient

class AccountUsernameMethodHandler : MethodHandler() {
    override fun execute(
        client: MobileAuthenticationClient,
        channelInMessage: ChannelInMessage
    ) {
        val message = validate<AccountUsernameMessage>(channelInMessage)
        val operation = OperationCache.getOperation<UserInteractionOperation>(message.operationId)
        val state = operation.state as? SelectAccount
            ?: throw IllegalStateException("UserInteractionState is missing or not SelectAccount for operation!")
        state.accountSelectionHandler?.username(message.username)
        MethodChannelHandler.resolve(ReactMethodName.ACCOUNT_USERNAME, null)
    }
}

File: ios/Methods/UserInteraction/AccountUsernameMethodHandler.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

struct AccountUsernameMethodHandler: MethodHandler {
	func execute(using _: MobileAuthenticationClient, with message: ChannelInMessage) throws {
		let message: AccountUsernameMessage = try validate(message: message)
		let operation: UserInteractionOperation = try OperationCache.shared.read(by: message.operationId)
		guard let state = operation.state as? SelectAccountState else {
			throw PluginError.illegalState("Select Account state is expected!")
		}

		state.handler?.username(message.username)
	}
}
